File: Swift-C-Interop/cfunctions.swift
// ------------------------------------------------------------------------------
// Copyright by Uwe Arzt mailto:mail@uwe-arzt.de, https://uwe-arzt.de
// under BSD License, see https://uwe-arzt.de/bsd-license
// ------------------------------------------------------------------------------

import Foundation

@_silgen_name("ret_uint8") func ret_uint8(_ i: UInt8) -> UInt8
@_silgen_name("ret_int8") func ret_int8(_ i: Int8) -> Int8
@_silgen_name("ret_uint64") func ret_uint64(_ i: UInt64) -> UInt64
@_silgen_name("ret_int64") func ret_int64(_ i: Int64) -> Int64

@_silgen_name("ret_float") func ret_float(_ f: Float32) -> Float32
@_silgen_name("ret_double") func ret_double(_ f: Double) -> Double

@_silgen_name("ret_buf") func ret_buf(_ buf: UnsafePointer<CChar>, len: size_t) -> UnsafeMutablePointer<CChar>

enum test_enum_t : Int {
    case en_one = 1
    case en_two = 2
}

@_silgen_name("ret_enum") func ret_enum(_ en: Int) -> Int

@_silgen_name("call") func call(_ cb: @convention(c) (Int) -> Int, i: Int) -> Int

File: Swift-C-Interop/cfunctions.c
// ------------------------------------------------------------------------------
// Copyright by Uwe Arzt mailto:mail@uwe-arzt.de, https://uwe-arzt.de
// under BSD License, see https://uwe-arzt.de/bsd-license
// ------------------------------------------------------------------------------

#include "cfunctions.h"

uint8_t ret_uint8(uint8_t i) { return i; }
int8_t ret_int8(int8_t i) { return i; }
uint64_t ret_uint64(uint64_t i) { return i; }
int64_t ret_int64(int64_t i) { return i; }

float ret_float(float f) { return f; }
double ret_double(double f) { return f; }

char* ret_buf(const char* buf, size_t len) {
    return (char*)buf;
}

test_enum_t ret_enum(test_enum_t en) {
    return en;
}

int call(callback_ptr cb, int i) {
    return cb(i);
}
